### PR TITLE
lxd/instance/qemu: Allow using external firmware or kernel

### PR DESCRIPTION
**Pull request by stgraber**
With this change, the -bios or -kernel options inside of raw.qemu will
cause LXD to log a warning message and then proceed to disable the
normal firmware setup.

This allows for booting LXD off of an external firmware file or directly
into a specific kernel. Such firmware or kernel will likely need a
matching raw.apparmor entry so that QEMU can actually access them.

Signed-off-by: Stéphane Graber <stgraber@ubuntu.com>

### CODE REVIEW

File: lxd/instance/drivers/driver_qemu.go [hunk 1]
```diff
@@ -2480,13 +2480,27 @@ func (d *qemu) generateQemuConfigFile(mountInfo *storagePools.MountInfo, busName
...
+	if d.expandedConfig["raw.qemu"] != "" {
+		rawOptions, err = shellquote.Split(d.expandedConfig["raw.qemu"])
+		if err != nil {
+			return "", nil, err
+		}
+	}
+
+	// Allow disabling the UEFI firmware.
+	if shared.StringInSlice("-bios", rawOptions) || shared.StringInSlice("-kernel", rawOptions) {
+		d.logger.Warn("Starting VM without standard firmware (-bios or -kernel was provided)")
```
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:2494]: "... provided in raw.qemu" for clarity?

**Comment by tomponline** [comment]
@stgraber also as this option is now "supported" (i.e recognized and specifically handled) should we document how to use `raw.qemu` to specify a particular bios or kernel?

CC @ru-fu

**Comment by stgraber** [comment]
There's a difference between tolerated and supported ;)
I don't think we should do anything that would encourage the use of `raw,qenu` by more users so I think I'm happy with the current level of documentation for it.

It's similar to `raw.lxc` where we do special treatment for some options too and take care to apply it at a spot where it's most likely to work but similarly don't have much more than one line in our docs about it.

**Comment by tomponline** [comment]
> 

OK, im thinking we should add tests for it at least (if feasible) so we detect it being broken?

**Pull request merged**